### PR TITLE
fix #7955 fully specialized generic failed at array construction

### LINKED ISSUE #7955: fully specialized generic failed at array construction
```Nim
type
  SBase[T, V] = ref object of RootObj
    val: T
    color: V
  SRC = ref object of SBase[string, int]
  SRD = ref object of SBase[string, int]

var x = [a, b, c] #ok
var y = [b, c, a] #failed
var z = [c, b, a] #failed
```

what need to be included in the test:
 * the above case
 * secondary form of ref -> e.g. from new(ref someobject)
 * ptr form, and it's secondary form
 * ordinary object -> perhaps generate runtime error because of object slicing

see the evolution of this bug in #7601, #7818, and #7906 

note: if possible, uncomment test at `block test_t4799_6` of #4799

### PR DESCRIPTION
**Pull request by bung87**
fix https://github.com/nim-lang/Nim/issues/7955

refs: https://github.com/nim-lang/Nim/issues/9136
handle case 
``` nim
type
  SBase[T, V] = ref object of RootObj
    val: T
    color: V
  SRC = ref object of SBase[string, int]

```
```
f:SBase[system.string, system.int]
a:SRC:ObjectType
```

related: https://github.com/nim-lang/Nim/issues/7956 (test case covered by t4799_2.nim t4799_3.nim )

### CODE REVIEW

File: compiler/types.nim [hunk 1]
```diff
@@ -1280,6 +1280,8 @@ proc commonSuperclass*(a, b: PType): PType =
   var ancestors = initIntSet()
   while x != nil:
     x = skipTypes(x, skipPtrs)
+    if skipRoot and sfCompilerProc in x.sym.flags:
+      break
```
Comment (Araq) [inline compiler/types.nim:1377]: What does this really skip here?

Comment (bung87) [reply compiler/types.nim:1377]: skip Rootobj and Exception they are compiler defined type

**Comment by beef331** [comment]
Is this not yet another "Not a bug, but expected behaviour"? `[b, c]` is an array of `typeof(b)` as such adding `c` to it is invalid. One should do:
```nim
type
  SBase[T, V] = ref object of RootObj
    val: T
    color: V
  SRC = ref object of SBase[string, int]
  SRD = ref object of SBase[string, int]

var a = SBase[string, int](val: "base", color: 1)
var b = SRC(val: "rc", color: 2)
var c = SRD(val: "rd", color: 3)

static:
  assert compiles([a, b, c])
  assert compiles([SBase[string, int](b), c])
  assert compiles([SBase[string, int](c), b])
  ```

File: compiler/types.nim (hunk 1 above)
Comment (Araq) [reply compiler/types.nim:1377]: Why would these be special cases?

Comment (bung87) [reply compiler/types.nim:1377]: make RootObj the common type, then next element don't match, the error message will be expect Root Obj but get X, sure A B inherit from RootObj directly they are common in RootObj, but no relationship in typeRel

Comment (Araq) [reply compiler/types.nim:1377]: Hmm ok.

**Pull request closed without merging**